File: clinical/src/main/java/com/hartwig/actin/clinical/feed/vitalfunction/VitalFunctionEntry.java
package com.hartwig.actin.clinical.feed.vitalfunction;

import java.time.LocalDate;

import com.hartwig.actin.clinical.feed.FeedEntry;

import org.immutables.value.Value;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

@Value.Immutable
@Value.Style(passAnnotations = { NotNull.class, Nullable.class })
public abstract class VitalFunctionEntry implements FeedEntry {

    @NotNull
    @Override
    public abstract String subject();

    @NotNull
    public abstract LocalDate effectiveDateTime();

    @NotNull
    public abstract String codeCodeOriginal();

    @NotNull
    public abstract String codeDisplayOriginal();

    @Nullable
    public abstract LocalDate issued();

    @NotNull
    public abstract String valueString();

    @NotNull
    public abstract String componentCodeCode();

    @NotNull
    public abstract String componentCodeDisplay();

    @NotNull
    public abstract String quantityUnit();

    public abstract double quantityValue();

}

File: clinical/src/main/kotlin/com/hartwig/actin/clinical/feed/vitalfunction/VitalFunctionEntry.kt
package com.hartwig.actin.clinical.feed.vitalfunction

import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.hartwig.actin.clinical.feed.EuropeanDecimalDeserializer
import com.hartwig.actin.clinical.feed.FeedEntry
import com.hartwig.actin.clinical.feed.FeedValidation
import com.hartwig.actin.clinical.feed.FeedValidationWarning
import com.hartwig.actin.clinical.feed.FeedValidator
import com.hartwig.actin.clinical.feed.TsvRow
import java.time.LocalDateTime

@TsvRow
data class VitalFunctionEntry(
    @JsonProperty("subject")
    override val subject: String,

    @JsonProperty("effectiveDateTime")
    val effectiveDateTime: LocalDateTime,

    @JsonProperty("code_display_original")
    val codeDisplayOriginal: String,

    @JsonProperty("component_code_display")
    val componentCodeDisplay: String,

    @JsonProperty("quantity_unit")
    val quantityUnit: String,

    @JsonProperty("value_quantity")
    @JsonDeserialize(using = EuropeanDecimalDeserializer::class)
    val quantityValue: Double?
) : FeedEntry {
    fun isValid(): Boolean {
        return true
    }
}

class VitalFunctionFeedValidator : FeedValidator<VitalFunctionEntry> {
    override fun validate(feed: VitalFunctionEntry): FeedValidation {
        val emptyCodeDisplayValidation =
            if (feed.codeDisplayOriginal.isEmpty()) listOf(
                FeedValidationWarning(
                    feed.subject,
                    "Empty vital function category"
                )
            ) else emptyList()
        val emptyQuantityValidation =
            if (feed.quantityValue == null) listOf(FeedValidationWarning(feed.subject, "Empty vital function value"))
            else emptyList()
        val noCategoryValidation =
            if (feed.codeDisplayOriginal.isNotEmpty() && VitalFunctionExtraction.toCategory(feed.codeDisplayOriginal) == null) listOf(
                FeedValidationWarning(
                    feed.subject,
                    "Invalid vital function category"
                )
            ) else emptyList()

        val warnings = emptyCodeDisplayValidation + emptyQuantityValidation + noCategoryValidation
        return FeedValidation(warnings.isEmpty(), warnings)
    }
}